### DISCUSSION TITLE
How to use group of aggrid?

### DISCUSSION THREAD

**Discussion by tz301** [zauberzeug/nicegui#3182] [Q&A]
### Question

From docs of aggrid: https://www.ag-grid.com/angular-data-grid/grouping-opening-groups/

We can set group of aggrid as below:

![image](https://github.com/zauberzeug/nicegui/assets/35760938/04dd7126-d14e-43af-895a-0f5620333f7f)

I have tried, but not working.

```py
grid = ui.aggrid({
    'columnDefs': [
        {'headerName': 'Name', 'field': 'name'},
        {'headerName': 'Age', 'field': 'age'},
        {'headerName': 'Gender', 'field': 'gender', 'rowGroup': True},
    ],
    'rowData': [
        {'name': 'Alice', 'age': 18, 'gender': 'male'},
        {'name': 'Bob', 'age': 21, 'gender': 'male'},
        {'name': 'Jason', 'age': 28, 'gender': 'female'},
    ],
    'groupDefaultExpanded': True
}, theme='alpine')
```

![image](https://github.com/zauberzeug/nicegui/assets/35760938/8e2f852f-f5c5-4c08-be07-0cdecef50021)

**Comment by falkoschindler** [comment]
Hi @tz301,

Unfortunately, NiceGUI doesn't support AG Grid Enterprise at the moment, which is required for row grouping.
See feature request #2431.

**Comment by tz301** [comment]
Hi @falkoschindler ,

Since I only need one-level grouping, I have tried to implement this by adding an extra column as below, a hided column "visibility" is used to control the grouping expand/collapse.

```py
grid = ui.aggrid({
    'columnDefs': [
        {'headerName': '', 'field': 'group', 'cellStyle': {'textAlign': 'right'}},
        {'headerName': 'Gender', 'field': 'gender'},
        {'headerName': 'Name', 'field': 'name'},
        {'headerName': 'Age', 'field': 'age'},
        {'headerName': 'Visibility', 'field': 'visibility', 'hide': True},
    ],
    'rowData': [
        {'group': '-', 'gender': 'male (2)', 'visibility': True},
        {'name': 'Alice', 'age': 18, 'gender': 'male', 'visibility': True},
        {'name': 'Bob', 'age': 21, 'gender': 'male',  'visibility': True},
        {'name': 'Jason', 'age': 28, 'gender': 'female', 'visibility': True},
    ],
    ':isExternalFilterPresent': '() => true',
    ':doesExternalFilterPass': '(node) => node.data.visibility == true'

}, theme='alpine')


def _on_cell_clicked(event):
    if grid.options['rowData'][0]['group'] == '-':
        grid.options['rowData'][0]['group'] = '+'
        grid.options['rowData'][1]['visibility'] = False
        grid.options['rowData'][2]['visibility'] = False
    else:
        grid.options['rowData'][0]['group'] = '-'
        grid.options['rowData'][1]['visibility'] = True
        grid.options['rowData'][2]['visibility'] = True
    grid.update()

grid.on('cellClicked', _on_cell_clicked)
```

![image](https://github.com/zauberzeug/nicegui/assets/35760938/456a6e1c-7c53-4bdb-beb7-68ef3c7bfb8a)
![image](https://github.com/zauberzeug/nicegui/assets/35760938/f1ee92d6-6c1b-4b57-a295-ed9d47d65143)

The behaviour seems working. However, when I clicked "+"/"-", the ui seems flash suddenly and then update.

I don't know if `grid.update` should behave like this. But I have tried demo on https://www.ag-grid.com/angular-data-grid/data-update-row-data/, the update seems very smooth.

**Reply by falkoschindler** [reply]
Currently we use `this.grid = new agGrid.Grid(this.$el, this.gridOptions);` to update an existing AG Grid. Maybe we can call `gridApi.setGridOption("rowData", rowData)` instead. But strangely I can't find this method in the AG Grid instance. I try to call it via `ui.run_javascript` or `grid.run_grid_method`, but JavaScript says "ReferenceError: setGridOption is not defined".

**Reply by tz301** [reply]
Hi @falkoschindler,

From https://www.ag-grid.com/javascript-data-grid/grid-api/#reference-gridOptions, it seems `setGridOption` or `updateGridOptions` can be used?

![image](https://github.com/zauberzeug/nicegui/assets/35760938/79772a35-fd60-4aed-ac0e-6388a5eefa95)

**Reply by falkoschindler** [reply]
Yes, that's what I thought. But this is me trying to find the method within the `grid` or `gridOptions` instance (element 4 being the grid from your example):

<img width="317" alt="Screenshot 2024-06-21 at 14 58 08" src="https://github.com/zauberzeug/nicegui/assets/5767091/e674f417-c307-4fd6-b2c9-42281c816f62">

**Reply by tz301** [reply]
Hi @falkoschindler,

Could you help to solve this such as search for help in Ag Grid github.

We have used nicegui for our product and got very good feedbacks. The update for aggrid may help to improve out UI.

Thanks!

**Reply by falkoschindler** [reply]
I looked into it once again and it seems to be a problem (or missing feature) in version 30.2.0. The following plain HTML/JavaScript example creates a grid and the function `grid.gridOptions.api.setGridOption` is defined, as can be checked using the JavaScript console:
```html
<html>
  <body>
    <div id="grid"></div>
    <script src="https://cdn.jsdelivr.net/npm/ag-grid-community/dist/ag-grid-community.min.js"></script>
    <script>
      document.addEventListener("DOMContentLoaded", () => {
        window.grid = new agGrid.Grid(document.querySelector("#grid"), {});
      });
    </script>
  </body>
</html>
```
But when using version 30.2.0, the function `grid.gridOptions.api.setGridOption` is undefined:
```html
    <script src="https://cdn.jsdelivr.net/npm/ag-grid-community@30.2.0/dist/ag-grid-community.min.js"></script>
```
So I suggest waiting for the next major release of [NiceGUI 2.0.0](https://github.com/zauberzeug/nicegui/milestone/82), where we will upgrade all JavaScript dependencies.

**Reply by tz301** [reply]
Excellent，I'll wait for it! Do you have approximate release date of it?

**Reply by falkoschindler** [reply]
We're planning to release by the end of August. 🤞🏻
